File: ios/Barry/Shared/Tendency.swift
//  Tendency.swift
//  Barry — Shared
//
//  Single source of truth for tendency thresholds on the Swift side. Mirrors
//  backend/app/tendency.py (brief §4.1). Keep the two in sync BY HAND — there is
//  intentionally no codegen; the table is small and stable.

import Foundation
import SwiftUI

/// 3-hour pressure tendency classification.
enum TendencyClass: String, Codable, CaseIterable {
    case risingFast = "rising_fast"
    case rising
    case steady
    case falling
    case fallingMod = "falling_mod"
    case fallingFast = "falling_fast"

    // Thresholds (hPa per 3 hours, signed; negative = falling).
    static let risingFastT = 1.5
    static let risingT = 0.5
    static let steadyT = -0.5
    static let fallingT = -1.5
    static let fallingModT = -3.0

    /// Classify a signed 3-hour delta. Mirrors `classify()` in tendency.py.
    static func classify(delta3h d: Double) -> TendencyClass {
        if d >= risingFastT { return .risingFast }
        if d >= risingT { return .rising }
        if d > steadyT { return .steady }
        if d > fallingT { return .falling }
        if d > fallingModT { return .fallingMod }
        return .fallingFast
    }

    var isFalling: Bool {
        switch self {
        case .falling, .fallingMod, .fallingFast: return true
        default: return false
        }
    }

    /// SF Symbol name for the trend glyph (brief §4.1 icon column).
    var symbolName: String {
        switch self {
        case .risingFast: return "arrow.up"
        case .rising: return "arrow.up.right"
        case .steady: return "arrow.right"
        case .falling: return "arrow.down.right"
        case .fallingMod: return "arrow.down"
        case .fallingFast: return "arrow.down.to.line"
        }
    }

    /// Short human label.
    var label: String {
        switch self {
        case .risingFast: return "Rising fast"
        case .rising: return "Rising"
        case .steady: return "Steady"
        case .falling: return "Falling"
        case .fallingMod: return "Falling steadily"
        case .fallingFast: return "Falling sharply"
        }
    }
}

/// Continuous color-intensity model. Mirrors `intensity()` in tendency.py.
enum TendencyIntensity {
    // Band: |d| from FLOOR..CEIL hPa mapped to 0..1 (pale amber -> deep red).
    // See the note in tendency.py: set floor to 0.0 to match the brief's example
    // JSON instead of its prose.
    static let floor = 1.5
    static let ceil = 4.0

    static func intensity(delta3h d: Double) -> Double {
        let mag = abs(d)
        let raw = (mag - floor) / (ceil - floor)
        return min(1.0, max(0.0, raw))
    }
}

extension TendencyClass {
    /// Color for the glyph / complication. Falling classes blend amber -> red by
    /// `intensity`; rising/steady are fixed. Intensity drives *depth*, not hue,
    /// so the complication conveys magnitude at a glance (brief §4.1).
    func color(intensity: Double) -> Color {
        switch self {
        case .risingFast: return Color.green
        case .rising: return Color(red: 0.55, green: 0.85, blue: 0.45) // light green
        case .steady: return Color.gray
        case .falling, .fallingMod, .fallingFast:
            return Self.amberToRed(intensity)
        }
    }

    /// Linear blend pale-amber -> deep-red over t in 0...1.
    static func amberToRed(_ t: Double) -> Color {
        let amber = (r: 0.98, g: 0.75, b: 0.30)
        let red = (r: 0.78, g: 0.10, b: 0.10)
        let c = min(1.0, max(0.0, t))
        return Color(
            red: amber.r + (red.r - amber.r) * c,
            green: amber.g + (red.g - amber.g) * c,
            blue: amber.b + (red.b - amber.b) * c
        )
    }
}

File: backend/app/tendency.py
"""3-hour pressure tendency: classification + continuous intensity.

This is the single source of truth for tendency thresholds on the Python side.
The Swift apps mirror these constants in Shared/Tendency.swift — keep the two in
sync by hand (there is intentionally no codegen; the table is small and stable).

See the project brief §4.1 for the canonical table.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Optional, Sequence

# --- Classification thresholds (hPa per 3 hours, signed; negative = falling) ---
#
#   d >= +1.5          rising_fast    green
#   +0.5 <= d < +1.5   rising         light green
#   -0.5 <  d < +0.5   steady         neutral/gray
#   -1.5 <  d <= -0.5  falling        pale amber
#   -3.0 <  d <= -1.5  falling_mod    amber
#   d <= -3.0          falling_fast   deep red
RISING_FAST = 1.5
RISING = 0.5
STEADY = -0.5
FALLING = -1.5
FALLING_MOD = -3.0

# Intensity gradient band: |d| is mapped from FALL_MIN -> FALL_MAX onto 0.0 -> 1.0
# (pale amber -> deep red), clamped outside that range. At the falling threshold
# (1.5 hPa) intensity is 0.0; at 4.0 hPa and beyond it is 1.0. This matches the
# brief's "pale amber at -1.5 -> deep red at -4+" intent.
#
# NOTE: the brief's example JSON shows delta3h -2.4 -> intensity 0.6, which is
# |d|/4.0 rather than this banded mapping ((2.4-1.5)/2.5 = 0.36). We use the
# banded mapping because it expresses the stated design intent (0 at the band
# floor, 1 at the band ceiling) and gives the complication a usable dynamic range
# across the falling classes. If you'd rather match the example literally, set
# INTENSITY_FLOOR = 0.0.
INTENSITY_FLOOR = 1.5
INTENSITY_CEIL = 4.0


@dataclass(frozen=True)
class Tendency:
    delta3h: float
    cls: str  # "class" is reserved; serialized as "class" in the API layer
    intensity: float


def classify(delta3h: float) -> str:
    """Map a signed 3-hour delta (hPa) to a tendency class."""
    d = delta3h
    if d >= RISING_FAST:
        return "rising_fast"
    if d >= RISING:
        return "rising"
    if d > STEADY:
        return "steady"
    if d > FALLING:
        return "falling"
    if d > FALLING_MOD:
        return "falling_mod"
    return "falling_fast"


def intensity(delta3h: float) -> float:
    """Continuous 0.0-1.0 magnitude for color intensity, based on |delta3h|.

    Mapped from INTENSITY_FLOOR..INTENSITY_CEIL hPa onto 0..1 and clamped.
    Returned for all classes; the complication only renders it for falling ones.
    """
    mag = abs(delta3h)
    span = INTENSITY_CEIL - INTENSITY_FLOOR
    raw = (mag - INTENSITY_FLOOR) / span
    return max(0.0, min(1.0, raw))


def tendency_from_delta(delta3h: float) -> Tendency:
    return Tendency(
        delta3h=round(delta3h, 2),
        cls=classify(delta3h),
        intensity=round(intensity(delta3h), 3),
    )


def delta3h_from_series(
    times: Sequence[datetime],
    values: Sequence[Optional[float]],
    *,
    now: Optional[datetime] = None,
    target_hours: float = 3.0,
    tolerance_hours: float = 1.5,
) -> Optional[float]:
    """Compute the 3-hour delta from a time series when presTend is unavailable.

    Returns (value at the most recent sample) - (value nearest ~3h before it).
    The earlier sample must fall within `tolerance_hours` of the target offset,
    otherwise we cannot honestly call it a 3-hour tendency and return None.

    `times`/`values` need not be sorted; None values are ignored.
    """
    pairs = [
        (t, v)
        for t, v in zip(times, values)
        if v is not None
    ]
    if len(pairs) < 2:
        return None
    pairs.sort(key=lambda p: p[0])

    latest_t, latest_v = pairs[-1]
    target_t = latest_t.timestamp() - target_hours * 3600.0

    best = None
    best_gap = None
    for t, v in pairs[:-1]:
        gap = abs(t.timestamp() - target_t)
        if best_gap is None or gap < best_gap:
            best_gap = gap
            best = (t, v)

    if best is None or best_gap is None:
        return None
    if best_gap > tolerance_hours * 3600.0:
        return None

    _, earlier_v = best
    return latest_v - earlier_v


def resolve_tendency(
    pres_tend: Optional[float],
    times: Sequence[datetime],
    values: Sequence[Optional[float]],
    *,
    now: Optional[datetime] = None,
) -> Optional[Tendency]:
    """Prefer the station-reported presTend; otherwise compute from the series.

    Returns None when neither source can produce an honest 3-hour delta.
    """
    if pres_tend is not None:
        return tendency_from_delta(pres_tend)
    delta = delta3h_from_series(times, values, now=now)
    if delta is None:
        return None
    return tendency_from_delta(delta)
